metaswitch/common/alarms.py

```diff
@@ -31,36 +31,281 @@
 # "OpenSSL Licenses" means the OpenSSL License and Original SSLeay License
 # under which the OpenSSL Project distributes the OpenSSL toolkit software,
 # as those licenses appear in the file LICENSE-OPENSSL.
+"""
+Interface for raising and clearing alarms in Python code.
 
+This module provides the interface for raising and clearing alarms in our
+python code.  Alarms should be raised using the classes in this module to
+ensure that they have the correct retry and re-synchronization logic applied.
 
-# This module provides a method for transporting alarm requests to a net-
-# snmp alarm sub-agent for further handling. If the agent is unavailable,
-# the request will timeout after 2 seconds and be dropped.
-
+Users should use the alarm_manager to get an instance of the alarm they
+wish to raise, then use its set and clear methods to update alarm state.
 
+The module also exposes constants expressing alarm severities.
+"""
+# Note, the class structure is based on the alarm objects in cpp-common.
+# The two implementations should be kept in step where possible.
 import logging
+import atexit
+from monotonic import monotonic
+import threading
+import weakref
 import imp
+from alarm_severities import (CLEARED,
+                              INDETERMINATE,
+                              CRITICAL,
+                              MAJOR,
+                              MINOR,
+                              WARNING)
 
 _log = logging.getLogger(__name__)
 
-sendrequest = None
+# Imported sendrequest method set up in issue_alarm.
+_sendrequest = None
+
+# How often to re-sync alarms in seconds.
+RE_SYNC_INTERVAL = 30
+
+
+class _AlarmManager(threading.Thread):
+    """
+    Singleton iterface to alarm code.
+
+    Use the instance alarm_manager as the single entry point into alarm
+    handling code.
+
+    Keeps a record of all alarms and makes sure they are re-raised
+    evert RE_SYNC_INTERVAL seconds.
+    """
+
+    def __init__(self):
+        super(_AlarmManager, self).__init__()
```

Comment:
We use Python 2.7 as it comes as standard on Ubuntu. Python 3 has a whole bunch of improvements - this is just the tip of the iceberg.